runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,734 @@
+    if expand_snippet then
+      apply_snippet(completion_item, suffix)
+    end
+
+    local command = completion_item.command
+    if command then
+      client:_exec_cmd(command, { bufnr = bufnr }, nil, function()
+        vim.lsp.log.warn(
+          string.format(
+            'Language server `%s` does not support command `%s`. This command may require a client extension.',
+            client.name,
+            command.command
+          )
+        )
+      end)
+    end
+  end
+
+  if completion_item.additionalTextEdits and next(completion_item.additionalTextEdits) then
+    local suffix = clear_word()
+    lsp.util.apply_text_edits(completion_item.additionalTextEdits, bufnr, offset_encoding)
+    apply_snippet_and_command(suffix)
+  elseif resolve_provider and type(completion_item) == 'table' then
+    local changedtick = vim.b[bufnr].changedtick
+
+    --- @param result lsp.CompletionItem
+    client.request(ms.completionItem_resolve, completion_item, function(err, result)
+      if changedtick ~= vim.b[bufnr].changedtick then
+        return
+      end
+
+      local suffix = clear_word()
+      if err then
+        vim.notify_once(err.message, vim.log.levels.WARN)
+      elseif result and result.additionalTextEdits then
+        lsp.util.apply_text_edits(result.additionalTextEdits, bufnr, offset_encoding)
+        if result.command then
+          completion_item.command = result.command
+        end
+      end
+
+      apply_snippet_and_command(suffix)
+    end, bufnr)
+  else
+    local suffix = clear_word()
+    apply_snippet_and_command(suffix)
+  end
+end
+
+--- @class vim.lsp.completion.BufferOpts
+--- @field autotrigger? boolean Whether to trigger completion automatically. Default: false
+
+--- @param client_id integer
+---@param bufnr integer
+---@param opts vim.lsp.completion.BufferOpts
+local function enable_completions(client_id, bufnr, opts)
+  if not buf_handles[bufnr] then
+    buf_handles[bufnr] = { clients = {}, triggers = {} }
+
+    -- Attach to buffer events.
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(_, buf)
+        buf_handles[buf] = nil
+      end,
+      on_reload = function(_, buf)
+        M.enable(true, client_id, buf, opts)
+      end,
+    })
+
+    -- Set up autocommands.
+    local group =
+      api.nvim_create_augroup(string.format('vim/lsp/completion-%d', bufnr), { clear = true })
+    api.nvim_create_autocmd('CompleteDone', {
+      group = group,
+      buffer = bufnr,
+      callback = function()
+        local reason = api.nvim_get_vvar('event').reason --- @type string
+        if reason == 'accept' then
+          on_complete_done()
+        end
+      end,
+    })
+    if opts.autotrigger then
+      api.nvim_create_autocmd('InsertCharPre', {
+        group = group,
+        buffer = bufnr,
+        callback = function()
+          on_insert_char_pre(buf_handles[bufnr])
+        end,
+      })
+      api.nvim_create_autocmd('InsertLeave', {
```

Comment:
Relying on InsertLeave can be wrong, because the user can exit the insert mode with ctrl-c, which doesn't trigger InsertLeave. Example: #29581. I noticed some other runtime files use InsertLeave, and they seem to be mostly fine. But I'm not quite sure if this module's use is correct.

Reply:
Rely on TextChangedI will be a better idea